presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java

```diff
@@ -50,6 +51,9 @@
     private final Map<Expression, Symbol> expressionToSymbols = new HashMap<>();
     private final Map<Expression, Expression> expressionToExpressions = new HashMap<>();
 
+    // memoize translated expressions so that the same output expression will be produced for each input expression
+    private final IdentityHashMap<Expression, Expression> translatedExpressions = new IdentityHashMap<>();
```

Comment:
I think that to make this work you would need to make `TranslationMap` to be immutable. Then you could have `CachingTranslationMap` around it. Alternatively, you could clean that map every time somebody calls any of `put*` method, but this kind of approach is more error-prone to me.